### PR TITLE
Account for Timezone when Casting Timestamp to Date32

### LINKED ISSUE #5598: confused result when cast timestamp with timezone to date32
**Which part is this question about**

cast from timestamp with timezone to date32

Related issues

I modify the `test_cast_timestamp_to_date32` UT to get the result of 

https://github.com/apache/arrow-datafusion/issues/9982

The  result of 
```
        let array =
            TimestampMillisecondArray::from(vec![Some(25201000), Some(111599000), None])
                .with_timezone("-07:00".to_string());
```

is 

```
PrimitiveArray<Timestamp(Millisecond, Some("-07:00"))>
[
  1970-01-01T00:00:01-07:00,
  **1970-01-01T23:59:59-07:00,**
  null,
]
```

the second value of timestamp is `1970-01-01T23:59:59-07:00`, but the result of `date32` is `1970-01-02`.

It's confused with me, and also is different from other engine.

**Describe your question**

**Additional context**

### PR DESCRIPTION
**Pull request by Lordworms**
# Which issue does this PR close?

Closes #5598

# Rationale for this change
 consider time zone when casting timestamp -> Date32

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-array/src/array/primitive_array.rs [hunk 1]
```diff
@@ -1366,7 +1366,13 @@ impl<T: ArrowTimestampType> PrimitiveArray<T> {
 
     /// Construct a timestamp array with new timezone
     pub fn with_timezone(self, timezone: impl Into<Arc<str>>) -> Self {
-        self.with_timezone_opt(Some(timezone.into()))
+        let timezone_str = timezone.into().to_lowercase();
+
+        if timezone_str == "utc" {
```
Comment (tustvold) [inline arrow-array/src/array/primitive_array.rs:1371]: Is this necessary?

File: arrow-array/src/timezone.rs [hunk 2]
```diff
@@ -257,7 +257,12 @@ mod private {
     /// An Arrow [`TimeZone`]
     #[derive(Debug, Copy, Clone)]
     pub struct Tz(FixedOffset);
-
+    impl Tz {
+        /// get timezone
+        pub fn get_time_zone(&self) -> FixedOffset {
```
Comment (tustvold) [inline arrow-array/src/timezone.rs:262]: This will misbehave if the chrono-tz feature is enabled and we support timezones with DST

**Review by tustvold — COMMENTED** [review]
I think this needs to make use of chrono, i.e. using `DateTime<Tz>` in order to be correct

File: arrow-array/src/array/primitive_array.rs (hunk 1 above)
Comment (Lordworms) [reply arrow-array/src/array/primitive_array.rs:1371]: I need this to help transfer timezone str to corresponding Tz

File: arrow-array/src/timezone.rs (hunk 2 above)
Comment (Lordworms) [reply arrow-array/src/timezone.rs:262]: Sure, I'll change it

File: arrow-array/src/array/primitive_array.rs [hunk 3]
```diff
@@ -1366,7 +1366,12 @@ impl<T: ArrowTimestampType> PrimitiveArray<T> {
 
     /// Construct a timestamp array with new timezone
     pub fn with_timezone(self, timezone: impl Into<Arc<str>>) -> Self {
-        self.with_timezone_opt(Some(timezone.into()))
+        let timezone_arc = timezone.into();
+        if timezone_arc.eq_ignore_ascii_case("utc") {
```
Comment (tustvold) [inline arrow-array/src/array/primitive_array.rs:1370]: This should be unnecessary

File: arrow-array/src/timezone.rs [hunk 4]
```diff
@@ -83,7 +83,21 @@ mod private {
...
     }
-
+    use chrono::DateTime;
+    use chrono::Utc;
+    impl Tz {
+        /// get timezone
+        pub fn get_time_zone_min(&self) -> i32 {
+            match self.0 {
+                TzInner::Timezone(tz) => {
+                    let utc_datetime: DateTime<Utc> = Utc::now();
```
Comment (tustvold) [inline arrow-array/src/timezone.rs:93]: This is incorrect, the timezone must be computed with reference to the time in question. Perhaps you could look at the other temporal cast kernels for inspiration

**Review by tustvold — CHANGES_REQUESTED** [review]

**Review by liukun4515 — CHANGES_REQUESTED** [review]
I think the fix is not right, we should not follow the `timezone` in the datatype to convert the value of date32.

We need to add the timezone config in the `CastOptions` which can't be configured by other system like datafusion.

**Comment by liukun4515** [comment]
The value of timezone in the datatype of `timestamp` means the value of timestamp is a reference to the UTC/UNIX time.
https://github.com/apache/arrow/blob/main/format/Schema.fbs#L283

**Comment by tustvold** [comment]
> We need to add the timezone config in the CastOptions which can't be configured by other system like datafusion.

I think the conversion should take the date of the timestamp in the given timezone, we should not need to configure this on `CastOptions`.

So for example if we had `2019-10-12T07:20:50.52 +10:00`, the date returned would be `2019-10-12` even though with reference to the UTC epoch (which is what the timestamp integer represents) the date would be `2019-10-11`

File: arrow-array/src/timezone.rs (hunk 4 above)
Comment (Lordworms) [reply arrow-array/src/timezone.rs:93]: Sure, I was just looking for a function to convert the timezone to int32/64

**Comment by Lordworms** [comment]
@tustvold Hi, I have refactored the code to use DateTime, please take a look when you have time. Thank you!

**Review by tustvold — APPROVED** [review]
Thank you for this, I took the liberty of refactoring it slightly and adding a little more test coverage. Nicely done

**Comment by liukun4515** [comment]
> > We need to add the timezone config in the CastOptions which can't be configured by other system like datafusion.
> 
> The conversion should take the date of the timestamp in the timezone configured on the timestamp data type, we should not need to configure this on `CastOptions`, we have very carefully avoided introducing the notion of a "local" timezone.
> 
> So for example if we had `2019-10-12T07:20:50.52 +10:00`, the date returned would be `2019-10-12` even though with reference to the UTC epoch (which is what the timestamp integer encodes) the date would be `2019-10-11`.



Hi  @tustvold 
From the definition of the timestamp in the arrow schema: https://github.com/apache/arrow/blob/main/format/Schema.fbs#L283

It's will cause inconvenient usage, let me give a example:

When I write a parquet file with timestamp column using the spark engine, the timestamp column is encoded by https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#instant-semantics-timestamps-normalized-to-utc and is normalized to UTC.

But when I use arrow-parquet to read the timestamp column, i will just get the column with timestamp(unit, "UTC") data type by this code https://github.com/apache/arrow-rs/blob/0d031cc8aa81296cb1bdfedea7a7cb4ec6aa54ea/parquet/src/arrow/schema/primitive.rs#L86 and https://github.com/apache/arrow-rs/blob/0d031cc8aa81296cb1bdfedea7a7cb4ec6aa54ea/parquet/src/arrow/schema/primitive.rs#L232

It's even possible to use `int96` to store the timestamp in the parquet file, we will get the column with timestamp(nanosecond, None) data type by https://github.com/apache/arrow-rs/blob/0d031cc8aa81296cb1bdfedea7a7cb4ec6aa54ea/parquet/src/arrow/schema/primitive.rs#L99

But i need this is used in the  `UTC-7` timezone, how did you resolve this issue? @alamb

**Comment by tustvold** [comment]
You will need to explicitly cast the timezone back, unfortunately as parquet has no mechanism for encoding the timezone this will need to be done manually. Arrow writers embed the arrow schema allowing this to be done automatically, but Spark doesn't do this

**Comment by alamb** [comment]
Maybe we can add some way to the parquet arrow reader to override its choice of data type for certain columns to allow users to specify types for cases where it is not clear from the parquet file itself. 

@mapleFU and I have been discussing the need for something similar for deciding what Array type to use when reading strings from Parquet files on https://github.com/apache/arrow-rs/issues/5530 -- see https://github.com/apache/arrow-rs/issues/5530#issuecomment-2052223254)

If we had such an API then people using spark created parquet files could specify that the timestamp column should always be UTC (as suggested by @tustvold ) without having to add an explicit cast afterwards

**Comment by liukun4515** [comment]
> Maybe we can add some way to the parquet arrow reader to override its choice of data type for certain columns to allow users to specify types for cases where it is not clear from the parquet file itself.
> 
> @mapleFU and I have been discussing the need for something similar for deciding what Array type to use when reading strings from Parquet files on #5530 -- see [#5530 (comment)](https://github.com/apache/arrow-rs/issues/5530#issuecomment-2052223254))
> 
> If we had such an API then people using spark created parquet files could specify that the timestamp column should always be UTC (as suggested by @tustvold ) without having to add an explicit cast afterwards

I think it's a solution adding the option in the parquet reader, it can help us to resolve some issues.

But many issues can't be resolve smoothly.

As describe in the comment: https://github.com/apache/arrow-datafusion/issues/9981#issuecomment-2058149468
The same timestamp column maybe used by different timezone user, they may want to do compute on the same timestamp with different timezone.  How to do that? @tustvold Do we need to add or wrap the `cast` expr explicitly to the target timestamp column?

**Comment by tustvold** [comment]
> Do we need to add or wrap the cast expr explicitly to the target timestamp column?

Yes arrow does not have a notion of "local" timezone. Theoretically a DF frontend could add such a notion, but I would strongly recommend against it as it is leads to all sorts of peculiar bugs.

**Comment by liukun4515** [comment]
> > Do we need to add or wrap the cast expr explicitly to the target timestamp column?
> 
> Yes arrow does not have a notion of "local" timezone, users need to be explicit about the timezone they wish to use for a particular array. Theoretically a DF frontend could add such a notion, but I would strongly recommend against it as it is leads to all sorts of peculiar bugs.
> 
> The following would be an example DataFusion SQL query that is explicit about timezones
> 
> ```
> SELECT arrow_cast(time_column, 'Timestamp(Nanosecond, Some("+07:00"))';
> ```

Thanks for your feedback for 
> > Do we need to add or wrap the cast expr explicitly to the target timestamp column?
I also know that adding explicit `cast` expr is not a good way.

Maybe the suggestion from @alamb is a good solution to resolve this after thinking about it again.
We can specify the `timezone` which is defined from `parquet reader options` when read the timestamp column.
what do you think about this method?

**Comment by tustvold** [comment]
Filed #5657

**Pull request merged**